### ISSUE TITLE
ILMerge loses SourceLink information

### ISSUE THREAD

**Issue by thomaslevesque** [Microsoft/ILMerge#54]
I'm trying to use [SourceLink](https://github.com/dotnet/sourcelink) to enable source stepping. I can see that the original pdb contains SourceLink information (mapping from local directory to GitHub), but after ILMerge the new PDB no longer contains this information. Is there a way to preserve it?

**Comment by thomaslevesque** [comment]
Any news on this?

**Comment by mike-barnett** [comment]
Oops. Sorry! I hadn't seen this. Can you provide me with a  repro and then I can look at it?

**Comment by thomaslevesque** [comment]
Hi @mike-barnett,

I made a repro on [this branch](https://github.com/thomaslevesque/FakeItEasy/tree/sourcelink-ilmerge) of the FakeItEasy project

- clone the repo
- checkout the sourcelink-ilmerge branch
- cd to src/FakeItEasy
- dotnet build

If you look at the resulting pdb for net40 or net45, you'll see that it doesn't contain sourcelink information (I just open it in an hex editor and look for the string "github")
If you edit FakeItEasy.csproj, comment the ILMerge targets, and build again, sourcelink information is present in the pdb.

**Comment by mike-barnett** [comment]
Thanks! I can repro it. When ILMerge is *not* used I see just one occurrence of "github" in the pdb file. Is that correct? Do you know what element in the assembly that is attached to? That will make it easier for me to debug the problem.

**Comment by thomaslevesque** [comment]
> When ILMerge is _not_ used I see just one occurrence of "github" in the pdb file. Is that correct?

Yes. In fact, it's a JSON document which looks like this (reformatted):

```json
{
    "documents": {
        "C:\\Git\\FakeItEasy\\*": "https://raw.githubusercontent.com/thomaslevesque/FakeItEasy/589bd72d2f4d6c1b58afb286139519716512c8ec/*"
    }
}
```

Basically it's just a mapping from a local path prefix to an URI prefix.

> Do you know what element in the assembly that is attached to?

No... In fact, I'm not sure it's attached to any specific element of the assembly. I don't know much about the internals of a PDB... The maintainers of SourceLink might be able to help you with this.

https://github.com/dotnet/sourcelink

**Comment by mike-barnett** [comment]
Do you know if *any* debugging information in the PDB that ILMerge writes is valid? I.e., can you successfully debug the merged assembly? This may be an issue of the PDB file being the new format. Do you know if it is or not?

**Comment by thomaslevesque** [comment]
> Do you know if _any_ debugging information in the PDB that ILMerge writes is valid? I.e., can you successfully debug the merged assembly?

Yes, I can debug the merged assembly without issue, so the merged PDB must be valid.

> This may be an issue of the PDB file being the new format. Do you know if it is or not?

No, it not the new PDB format. We [explicitly use the old format](https://github.com/FakeItEasy/FakeItEasy/blob/23761e9655c213d954f0222b9706a73020dbcb6a/src/FakeItEasy/FakeItEasy.csproj#L36), because ILMerge doesn't support the new one.

**Comment by mike-barnett** [comment]
Ah, okay. I'll see if I can get in touch with the SourceLink folks and see where they're stashing that information in the PDB file.

**Comment by mike-barnett** [comment]
Okay, I've looked into it: in order to support SourceLink, ILMerge would have to stop using the old components for PDB reading and would have to start using the [Microsoft.DiaSymReader](https://www.nuget.org/packages/Microsoft.DiaSymReader/) package and new interfaces to read/write the PDB files. This is definitely too large a work item for me to do. Have you tried [il-repack](https://github.com/gluck/il-repack)? Does that support SourceLink?

I'm very sorry, but my time for working on ILMerge is extremely limited. If these kind of changes are going to be made, I would recommend doing it as part of il-repack or some other Mono.Cecil-based project.

**Comment by thomaslevesque** [comment]
> Have you tried [il-repack](https://github.com/gluck/il-repack)? Does that support SourceLink?

I'll check, but I don't think it does. Thanks for the suggestion.

I understand it's a pretty big undertaking. I'd like to help, but it would probably take more time than I have just to understand how ILMerge works...
